### PR TITLE
add omt mvn distribution

### PR DESCRIPTION
**Pull request by martinjankowiak**
multivariate normal with omt gradients. @fritzo, you can assume the backwards math is correct (there's a test!), but i'm mostly looking for CR for distributions interface etc. also i'm always confused about ctx/save_for_backward....

### CODE REVIEW

File: pyro/distributions/omt_mvn.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    :param torch.autograd.Variable loc: Mean.
+    :param torch.autograd.Variable scale_tril: Cholesky of Covariance matrix.
+    """
+    params = {"loc": constraints.real, "scale_tril": constraints.lower_triangular}
+    support = constraints.real
+    has_rsample = True
+    reparameterized = True
+
+    def __init__(self, loc, scale_tril):
+        self.loc = loc
```
Comment (fritzo) [inline pyro/distributions/omt_mvn.py]: It looks like you're inheriting from `MultivariateNormal` to get its `.log_prob` method. If this is the case, then you'll need to call its constructor
```py
def __init__(self, loc, scale_tril):
    covariance_matrix = torch.mm(scale_tril, scale_tril.t())
    super(OMTMultivariateNormal, self).__init__(loc, covariance_matrix)
    self.scale_tril = scale_tril
```

File: pyro/distributions/omt_mvn.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        self.scale_tril = scale_tril
+        assert(len(loc.size()) == 1), "OMTMultivariateNormal loc must be 1-dimensional"
+        assert(len(scale_tril.size()) == 2), "OMTMultivariateNormal scale_tril must be 2-dimensional"
+
+    @property
```
Comment (fritzo) [inline pyro/distributions/omt_mvn.py]: Remove this; it is provided by base class.

File: pyro/distributions/omt_mvn.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    def variance(self):
+        return self.covariance_matrix.diag()
+
+    @lazy_property
```
Comment (fritzo) [inline pyro/distributions/omt_mvn.py]: Remove this; the computation was already needed to create the base class.

Comment (martinjankowiak) [reply pyro/distributions/omt_mvn.py]: thanks! done

File: pyro/distributions/omt_mvn.py (hunk 2 above)
Comment (martinjankowiak) [reply pyro/distributions/omt_mvn.py]: thanks! done

File: pyro/distributions/omt_mvn.py (hunk 1 above)
Comment (martinjankowiak) [reply pyro/distributions/omt_mvn.py]: thanks! done

File: pyro/distributions/omt_mvn.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    :param torch.autograd.Variable loc: Mean.
+    :param torch.autograd.Variable scale_tril: Cholesky of Covariance matrix.
+    """
+    params = {"loc": constraints.real, "scale_tril": constraints.lower_triangular}
+    support = constraints.real
+    has_rsample = True
+    reparameterized = True
+
+    def __init__(self, loc, scale_tril):
+        assert(len(loc.size()) == 1), "OMTMultivariateNormal loc must be 1-dimensional"
```
Comment (fritzo) [inline pyro/distributions/omt_mvn.py]: Note you can use `x.dim()` instead of `len(x.size())`

**Comment by martinjankowiak** [comment]
added test_log_prob. do i need to add anything to pyro `__init__`s?

File: pyro/distributions/omt_mvn.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        assert(len(scale_tril.size()) == 2), "OMTMultivariateNormal scale_tril must be 2-dimensional"
+        covariance_matrix = torch.mm(scale_tril, scale_tril.t())
+        super(OMTMultivariateNormal, self).__init__(loc, covariance_matrix)
+        self.scale_tril = scale_tril
+
+    def rsample(self, sample_shape=torch.Size()):
+        return _OMTMVNSample.apply(self.loc, self.scale_tril, sample_shape + self.loc.shape)
+
+
+def sum_all_but_rightmost_dim(x):
```
Comment (fritzo) [inline pyro/distributions/omt_mvn.py]: Consider adding a generalized version to pyro/distributions/util.py
```py
def sum_leftmost(x, dim):
    if dim < 0:
        dim = x.dim() - dim
    if dim == 0:
        return x
    return x.contiguous().view(-1, *x.shape[dim:]).sum(0)

x = torch.ones(2,3,4)
assert sum_leftmost(x, -2).shape == (3, 4)
assert sum_leftmost(x, -1).shape == (4,)
```

File: pyro/distributions/omt_mvn.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    O(D^3), although the resulting gradient variance is generally expected to be lower.
+
+    A distribution over vectors in which all the elements have a joint Gaussian
+    density.
+
+    :param torch.autograd.Variable loc: Mean.
+    :param torch.autograd.Variable scale_tril: Cholesky of Covariance matrix.
+    """
+    params = {"loc": constraints.real, "scale_tril": constraints.lower_triangular}
+    support = constraints.real
```
Comment (fritzo) [inline pyro/distributions/omt_mvn.py]: Omit the `.support`, `.reparameterized`, and `.has_rsample` properties; these are already correct.

**Comment by fritzo** [comment]
> do i need to add anything to pyro __init__s?

Could you import this in pyro/distributions/__init__.py and add a section to docs/source/distributions.rst?

File: pyro/distributions/omt_mvn.py (hunk 5 above)
Comment (fritzo) [reply pyro/distributions/omt_mvn.py]: Actually this would be useful for me as well :smile:

Comment (martinjankowiak) [reply pyro/distributions/omt_mvn.py]: done, thanks

File: pyro/distributions/omt_mvn.py (hunk 6 above)
Comment (martinjankowiak) [reply pyro/distributions/omt_mvn.py]: done, thanks

File: pyro/distributions/omt_mvn.py (hunk 4 above)
Comment (martinjankowiak) [reply pyro/distributions/omt_mvn.py]: done, thanks

File: pyro/distributions/util.py [hunk 7]
```diff
@@ -98,6 +98,31 @@ def sum_rightmost(value, dim):
...
+    ```
+    x = torch.ones(2,3,4)
+    assert sum_leftmost(x, -2).shape == (3, 4)
+    assert sum_leftmost(x, -1).shape == (4,)
+    ```
+
+    :param torch.autograd.Variable x: A tensor
+    :param int dim: Specifies the number of dims to sum out (a non-positive integer)
+    """
+    assert(dim <= 0), "dim must be less than or equal to zero"
```
Comment (fritzo) [inline pyro/distributions/util.py]: Could you please implement the general interface I suggested.
```py
def sum_leftmost(x, dim):
    if dim < 0:
        dim += x.dim()
    ...
```
and add a test
```py
def test_sum_leftmost():
    x = torch.ones(2,3,4)
    assert sum_leftmost(x, 0).shape == (2, 3, 4)
    assert sum_leftmost(x, 1).shape == (3, 4)
    assert sum_leftmost(x, 2).shape == (4,)
    assert sum_leftmost(x, -1).shape == (4,)
    assert sum_leftmost(x, -2).shape == (3, 4)
```

Comment (martinjankowiak) [reply pyro/distributions/util.py]: i didn't understand your interface

Comment (martinjankowiak) [reply pyro/distributions/util.py]: since ``dim += x.dim()`` was different. i guess now i see it

Comment (fritzo) [reply pyro/distributions/util.py]: Sorry for being unclear. The interface should match `sum_rightmost()`: for positive integers `n`, `sum_leftmost(x, n)` sums out the `n` leftmost dimensions. For negative integers `-n`, `sum_leftmost(x, -n)` sums out all but the `n` rightmost dimensions, i.e. the leftmost `x.dim() - n` dimensions. This matches the behavior of negative indexing for many other operations.

Comment (fritzo) [reply pyro/distributions/util.py]: Thanks, I'm only requesting this because we'll be needing it in our work on `iarange` and parallel enumeration, so your efforts in this PR will pay off.

**Comment by martinjankowiak** [comment]
either way, tests are pretty thorough, would be hard to fool
